File: HW2/ProofSimplifier.java
import java.util.*;

public class ProofSimplifier {
    private List<ConvertedExpression> convertedHypothesisList;
    private List<ConvertedExpression> convertedExpressionList;

    private HashMap<ConvertedExpression, int[]> canMP;

    private HashMap<ConvertedExpression, Integer> convertedExpressionMap;

    private ConvertedExpression provableExpression;

    private List<Annotation> annotationList;

    private List<String> simplifiedExpressionStringList;

    public ProofSimplifier(String hypothesis, List<String> expressionList) {
        HypothesisConverter hypothesisConverter = new HypothesisConverter(hypothesis);

        convertedHypothesisList = hypothesisConverter.getHypothesisList();
        provableExpression = hypothesisConverter.getProvableExpression();

        convertedExpressionList = new ArrayList<>();

        for (String expression : expressionList) {
            ExpressionConverter expressionConverter = new ExpressionConverter(expression);

            convertedExpressionList.add(expressionConverter.getConvertedExpression());
        }

        simplifiedExpressionStringList = new ArrayList<>();
        annotationList = new ArrayList<>(Collections.nCopies(expressionList.size(), null)); //TODO: null

        convertedExpressionMap = new HashMap<>();
        canMP = new HashMap<>();

        simplify();
    }

    private void simplify() {
        checkAxiomAndHypothesis();

        for (int i = 0; i < convertedExpressionList.size(); i++) {
            ConvertedExpression expression = convertedExpressionList.get(i);

            if (!convertedExpressionMap.containsKey(expression) &&
                    canMP.containsKey(expression) &&
                    canMP.get(expression)[0] <= i) {

                annotationList.set(i, new Annotation(AnnotationType.MODUS_PONENS, canMP.get(expression)[1], canMP.get(expression)[2]));

                convertedExpressionMap.put(expression, i);
            }

            if (!convertedExpressionMap.containsKey(expression)) {
                continue;
            } else {
                if (expression.getSubExpression().getType().equals(ExpressionPartType.IMPLICATION) &&
                    convertedExpressionMap.containsKey(expression.getSubConvertedExpressions()[0]) &&
                    !canMP.containsKey(expression.getSubConvertedExpressions()[1])) {

                    canMP.put(expression.getSubConvertedExpressions()[1], new int[]{
                            Math.max(i, convertedExpressionMap.get(expression.getSubConvertedExpressions()[0])),
                            i,
                            convertedExpressionMap.get(expression.getSubConvertedExpressions()[0])
                    });
                }
            }
        }

        if (!convertedExpressionMap.containsKey(provableExpression)) {
            simplifiedExpressionStringList.add("Proof is incorrect");

            return;
        }

        List<Boolean> needExpression = new ArrayList<>(Collections.nCopies(convertedExpressionList.size(), false));
        needExpression.set(convertedExpressionMap.get(provableExpression), true);

        for (int i = convertedExpressionList.size() - 1; i >= 0; i--) {
            if (!needExpression.get(i)) {
                continue;
            }

            if (annotationList.get(i).getAnnotationType().equals(AnnotationType.MODUS_PONENS)) {
                for (int j = 0; j < annotationList.get(i).getIndexes().length; j++) {
                    needExpression.set(annotationList.get(i).getIndexes()[j], true);
                }
            }
        }

        List<Integer> numeration = new ArrayList<>();
        int index = 1;

        for (int i = 0; i < convertedExpressionList.size(); i++) {
            if (needExpression.get(i)) {
                numeration.add(index++);
            } else {
                numeration.add(-1);
            }
        }



        for (int i = convertedExpressionList.size() - 1; i >= 0; i--) {
            if (!needExpression.get(i)) {
                continue;
            }

            String expression = String.format("[%d. %s ", numeration.get(i), annotationList.get(i).getAnnotationType().getStringValue());

            for (int j = 0; j < annotationList.get(i).getIndexes().length; j++) {
                expression +=
                        (annotationList.get(i).getAnnotationType().equals(AnnotationType.MODUS_PONENS) ?
                                numeration.get(annotationList.get(i).getIndexes()[j]) :
                                annotationList.get(i).getIndexes()[j] + 1) +
                        (j == annotationList.get(i).getIndexes().length - 1 ? "" : ", ");
            }

            expression += "] ";

            expression += convertedExpressionList.get(i).toNormalString();

            simplifiedExpressionStringList.add(expression);
        }

        String hypothesis = "";

        if (convertedHypothesisList != null) {
            for (int i = 0; i < convertedHypothesisList.size(); i++) {
                hypothesis += convertedHypothesisList.get(i).toNormalString() +
                        (i == convertedHypothesisList.size() - 1 ? " " : ", ");
            }
        }

        hypothesis += "|- " + provableExpression.toNormalString();

        simplifiedExpressionStringList.add(hypothesis);

        Collections.reverse(simplifiedExpressionStringList);
    }

    private void checkAxiomAndHypothesis() {
        for (int i = 0; i < convertedExpressionList.size(); i++) {
            ConvertedExpression expression = convertedExpressionList.get(i);

            if (convertedExpressionMap.containsKey(expression)) {
                if (i == convertedExpressionList.size() - 1) {
                    convertedExpressionMap.put(expression, i);

                    annotationList.set(i, annotationList.get(convertedExpressionMap.get(expression)));
                }

                continue;
            }

            boolean deduced = false;

            if (deduced) {
                convertedExpressionMap.put(expression, i);
                continue;
            }

            if (convertedHypothesisList != null) {
                for (int j = 0; j < convertedHypothesisList.size(); j++) {
                    if (expression.equals(convertedHypothesisList.get(j))) {
                        annotationList.set(i, new Annotation(AnnotationType.HYPOTHESIS, j));

                        deduced = true;
                        break;
                    }
                }
            }

            if (deduced) {
                convertedExpressionMap.put(expression, i);
                continue;
            }

            for (int j = 0; j < Axiom.AXIOMS_NUMBER; j++) {
                if (Axiom.check(j, expression)) {
                    annotationList.set(i, new Annotation(AnnotationType.AXIOM, j));

                    deduced = true;
                    break;
                }
            }

            if (deduced) {
                convertedExpressionMap.put(expression, i);
                continue;
            }
        }
    }

    public List<String> getSimplifiedExpressionStringList() {
        return simplifiedExpressionStringList;
    }
}

File: Lab/B/proof_simplifier.cpp
#include "proof_simplifier.h"


string hypothesisStringForProofSimplifier;

converted_expression* provableExpression;

vector<converted_expression*> convertedExpressionList(0);

unordered_map<string, int> firstAppearance;

unordered_map<string, int> convertedHypothesisMap;

unordered_map<string, vector<int>> canMP;

vector<annotation*> annotationList(0);

vector<int> indexesToAnnotation(0);

void simplify(const string &hypothesis, vector<string> expressionVector) {
    //BEGIN: PREPARE
    hypothesisStringForProofSimplifier = hypothesis;

    pair<vector<converted_expression*>, converted_expression*> hypothesisPair = convertHypothesis(hypothesis);

    provableExpression = hypothesisPair.second;

    int ind = -1;
    for (converted_expression *expression : hypothesisPair.first) {
        ind++;
        convertedHypothesisMap[expression->getStringValue()] = ind;
    }

    for (string expression : expressionVector) {
        convertedExpressionList.push_back(convertExpression(expression));
    }

    annotationList.resize(expressionVector.size(), nullptr);
    //END: PREPARE

    if (!(provableExpression->equals(convertedExpressionList[convertedExpressionList.size() - 1]))) {
        cout << "Proof is incorrect";

        return;
    }

    checkAxiomAndHypothesis();

    convertedHypothesisMap.clear();

    int first;
    int second;

    string firstString;
    string secondString;

    for (int i = 0; i < convertedExpressionList.size(); i++) {
        converted_expression *expression = convertedExpressionList[i];

        if (firstAppearance[expression->getStringValue()] != i) {
            annotationList[i] = annotationList[firstAppearance[expression->getStringValue()]];
        }

        if (canMP.count(expression->getStringValue()) != 0) {
            vector<int> fromCanMP = canMP[expression->getStringValue()];

            if (firstAppearance[expression->getStringValue()] == i &&
                canMP.count(expression->getStringValue()) != 0 &&
                fromCanMP[0] < i) {

                indexesToAnnotation.clear();
                indexesToAnnotation.push_back(fromCanMP[1]);
                indexesToAnnotation.push_back(fromCanMP[2]);

                annotationList[i] = new annotation(MODUS_PONENS,
                                                   indexesToAnnotation);

                canMP.erase(expression->getStringValue());
            }
        }

        if (firstAppearance[expression->getStringValue()] == i && annotationList[i] == nullptr) {
            cout << "Proof is incorrect";

            return;
        } else {
            auto *castExpression = dynamic_cast<binary_operation *>(expression);

            if (castExpression != nullptr && castExpression->operation == const_token::IMPLICATION) {
                firstString = castExpression->first->getStringValue();

                if (firstAppearance.count(firstString) != 0) {
                    first = firstAppearance[firstString];
                    int maxx = max(i, first);

                    secondString = castExpression->second->getStringValue();
                    if (firstAppearance.count(secondString) != 0) {
                        second = firstAppearance[secondString];

                        if (second > maxx) {
                            indexesToAnnotation.clear();
                            indexesToAnnotation.push_back(maxx);
                            indexesToAnnotation.push_back(i);
                            indexesToAnnotation.push_back(first);

                            canMP[secondString] = indexesToAnnotation;
                        }
                    }
                }
            }
        }
    }

    cout << hypothesisStringForProofSimplifier << "\n";

    vector<bool> needExpression(convertedExpressionList.size(), false);
    needExpression[firstAppearance[provableExpression->getStringValue()]] =  true;

    for (int i = (int) convertedExpressionList.size() - 1; i >= 0; i--) {
        if (!needExpression[i]) {
            continue;
        }

        if (annotationList[i]->getAnnotationType() == MODUS_PONENS) {
            for (int j = 0; j < annotationList[i]->getIndexes().size(); j++) {
                needExpression[annotationList[i]->getIndexes()[j]] = true;
            }
        }
    }

    vector<int> numeration(0);
    int index = 1;

    for (int i = 0; i < convertedExpressionList.size(); i++) {
        if (needExpression[i]) {
            numeration.push_back(index++);
        } else {
            numeration.push_back(-1);
        }
    }




    for (int i = 0; i < convertedExpressionList.size(); i++) {
        if (!needExpression[i]) {
            continue;
        }


        cout << ("[");
        cout << (numeration[i]);
        cout << (". ");
        cout << (annotationList[i]->getStringValue());
        cout << (" ");

        for (int j = 0; j < annotationList[i]->getIndexes().size(); j++) {
            cout << (
                    (annotationList[i]->getAnnotationType() == MODUS_PONENS ?
                     numeration[annotationList[i]->getIndexes()[j]] :
                     annotationList[i]->getIndexes()[j] + 1));
            cout << (j == annotationList[i]->getIndexes().size() - 1 ? "" : ", ");
        }

        cout << ("] ");

        cout << (convertedExpressionList[i]->getStringValue()) << "\n";
    }
}

void checkAxiomAndHypothesis() {
    for (int i = 0; i < convertedExpressionList.size(); i++) {
        converted_expression *expression = convertedExpressionList[i];

        if (firstAppearance.count(expression->getStringValue()) == 0) {
            firstAppearance[expression->getStringValue()] = i;
        }

        bool deduced = false;

        if (convertedHypothesisMap.count(expression->getStringValue()) != 0) {
            indexesToAnnotation.clear();
            indexesToAnnotation.push_back(convertedHypothesisMap[expression->getStringValue()]);

            annotationList[i] = new annotation(HYPOTHESIS, indexesToAnnotation);

            deduced = true;
        }

        if (deduced) {
            continue;
        }


        for (int j = 0; j < getAxiomNumber(); j++) {
            if (checkAxiom(expression, j)) {
                indexesToAnnotation.clear();
                indexesToAnnotation.push_back(j);

                annotationList[i] = new annotation(AXIOM, indexesToAnnotation);

                deduced = true;
                break;
            }
        }
    }
}